File: compiler/codegen/instructions.go
package codegen

import (
	"fmt"
	"strconv"
)

type (
	Instruction string
	Directive   string
)

const (
	freeLABEL  = "FREE"
	nOne       = "-1"
	one        = "1"
	nFour      = "-4"
	four       = "4"
	pThirtySix = "36"

	JMP Instruction = "jmp"
	JMR Instruction = "jmr"
	BNZ Instruction = "bnz"
	BGT Instruction = "bgt"
	BLT Instruction = "blt"
	BRZ Instruction = "brz"
	MOV Instruction = "mov"
	LDA Instruction = "lda"
	STR Instruction = "str"
	LDR Instruction = "ldr"
	STB Instruction = "stb"
	LDB Instruction = "ldb"
	ADD Instruction = "add"
	ADI Instruction = "adi"
	SUB Instruction = "sub"
	MUL Instruction = "mul"
	DIV Instruction = "div"
	AND Instruction = "and"
	OR  Instruction = "or"
	CMP Instruction = "cmp"
	TRP Instruction = "trp"

	INT Directive = ".int"
	BYT Directive = ".byt"

	R0 = "R0"
	R1 = "R1"
	R2 = "R2"
	R3 = "R3"
	R4 = "R4"
	R5 = "R5"
	R6 = "R6"
	R7 = "R7"
	FP = "FP"
	SP = "SP"
	PC = "PC"
	SL = "SL"
)

// keeps track of how many bytes of opcodes and data was written out
var (
	bytsAllocated = 0
)

func op(label string, instruction Instruction, op1 string, op2 string) string {
	bytsAllocated += 12
	return fmt.Sprintf("%s %s %s %s", label, instruction, op1, op2)
}

func opnl(instruction Instruction, op1 string, op2 string) string {
	return op("", instruction, op1, op2)
}

func writeFREE() string {
	bytsAllocated += 4
	return fmt.Sprintf("%s %s %s", freeLABEL, INT, strconv.Itoa(bytsAllocated))
}

func dir(label string, dir Directive, val string) string {
	if dir == BYT {
		bytsAllocated += 1
	} else {
		bytsAllocated += 4
	}
	return fmt.Sprintf("%s %s %s", label, dir, val)
}

func dirln(direc Directive, val string) string {
	return dir("", direc, val)
}

File: vm/haskellvm/src/Instructions.hs
module Instructions
  ( Instruction(..)
  , Directive(..)
  , getInstructionFromWord
  , getInstructionValue
  ) where

import Data.Int 

data Instruction
  = JMP
  | JMR
  | BNZ
  | BGT
  | BLT
  | BRZ
  | MOV
  | LDA
  | STRrl
  | LDRrl
  | STBrl
  | LDBrl
  | ADD
  | ADI
  | SUB
  | MUL
  | DIV
  | AND
  | OR
  | CMP
  | TRP
  | STRrr
  | LDRrr
  | STBrr
  | LDBrr
  | RUN
  | END
  | BLK
  | LCK
  | ULK
  deriving (Show, Eq)

data Directive
  = INT
  | BYT

getInstructionValue :: Instruction -> Int32
getInstructionValue instruction =
  case instruction of
    JMP -> 1
    JMR -> 2
    BNZ -> 3
    BGT -> 4
    BLT -> 5
    BRZ -> 6
    MOV -> 7
    LDA -> 8
    STRrl -> 9
    LDRrl -> 10
    STBrl -> 11
    LDBrl -> 12
    ADD -> 13
    ADI -> 14
    SUB -> 15
    MUL -> 16
    DIV -> 17
    AND -> 18
    OR -> 19
    CMP -> 20
    TRP -> 21
    STRrr -> 22
    LDRrr -> 23
    STBrr -> 24
    LDBrr -> 25
    RUN -> 26
    END -> 27
    BLK -> 28
    LCK -> 29
    ULK -> 30

getInstructionFromWord :: Int32 -> Maybe Instruction
getInstructionFromWord word =
  case word of
    1 -> Just JMP
    2 -> Just JMR
    3 -> Just BNZ
    4 -> Just BGT
    5 -> Just BLT
    6 -> Just BRZ
    7 -> Just MOV
    8 -> Just LDA
    9 -> Just STRrl
    10 -> Just LDRrl
    11 -> Just STBrl
    12 -> Just LDBrl
    13 -> Just ADD
    14 -> Just ADI
    15 -> Just SUB
    16 -> Just MUL
    17 -> Just DIV
    18 -> Just AND
    19 -> Just OR
    20 -> Just CMP
    21 -> Just TRP
    22 -> Just STRrr
    23 -> Just LDRrr
    24 -> Just STBrr
    25 -> Just LDBrr
    26 -> Just RUN
    27 -> Just END
    28 -> Just BLK
    29 -> Just LCK
    30 -> Just ULK
    _ -> Nothing
